lib/location/method_gnss.c

```diff
@@ -828,13 +879,103 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
 }
 
 #if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+static bool method_gnss_agps_expiry_process(const struct nrf_modem_gnss_agps_expiry *agps_expiry)
+{
+	uint16_t ephe_expiry_threshold;
+	uint8_t expired_ephes_min_count;
+	uint8_t expired_ephes = 0;
+	uint8_t expired_alms = 0;
+
+	memset(&agps_request, 0, sizeof(agps_request));
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
+		ephe_expiry_threshold = PGPS_EXPIRY_THRESHOLD;
+	} else {
+		ephe_expiry_threshold = AGPS_EXPIRY_THRESHOLD;
+	}
+
+	for (int i = 0; i < NRF_MODEM_GNSS_NUM_GPS_SATELLITES; i++) {
+		if (agps_expiry->ephe_expiry[i] <= ephe_expiry_threshold) {
+			expired_ephes++;
+		}
+
+		if (agps_expiry->alm_expiry[i] <= AGPS_EXPIRY_THRESHOLD) {
+			expired_alms++;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
+		expired_ephes_min_count = PGPS_EPHE_MIN_COUNT;
+	} else {
+		expired_ephes_min_count = AGPS_EPHE_MIN_COUNT;
+	}
+
+	if (expired_ephes >= expired_ephes_min_count) {
+		agps_request.sv_mask_ephe = 0xffffffff;
+	}
+
+	if (expired_alms >= AGPS_ALM_MIN_COUNT) {
+		agps_request.sv_mask_alm = 0xffffffff;
+	}
+
+	if (agps_expiry->utc_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST;
+	}
+
+	if (agps_expiry->klob_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST;
+	}
+
+	if (agps_expiry->data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST;
+	}
+
+	if (agps_expiry->integrity_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST;
+	}
+
+	if (agps_expiry->data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_POSITION_REQUEST;
+	}
+
+	LOG_DBG("Expired ephemerides: %d, almanacs: %d", expired_ephes, expired_alms);
+
+	LOG_DBG("Assistance request sv_mask_ephe: 0x%08x, sv_mask_alm: 0x%08x, data_flags: 0x%02x",
+		agps_request.sv_mask_ephe, agps_request.sv_mask_alm, agps_request.data_flags);
+
+	return agps_request.sv_mask_ephe != 0x0 ||
+	       agps_request.sv_mask_alm != 0x0 ||
+	       agps_request.data_flags != 0x0;
+}
+
 static void method_gnss_agps_req_work_fn(struct k_work *work)
 {
-	/* Start and stop GNSS to trigger NRF_MODEM_GNSS_EVT_AGPS_REQ event if assistance data
-	 * is needed.
-	 */
-	nrf_modem_gnss_start();
-	nrf_modem_gnss_stop();
+	int err;
+	struct nrf_modem_gnss_agps_expiry agps_expiry;
+
+	err = nrf_modem_gnss_agps_expiry_get(&agps_expiry);
+	if (err) {
+		if (err == -NRF_EOPNOTSUPP) {
+			LOG_DBG("nrf_modem_gnss_agps_expiry_get() not supported by the modem "
+				"firmware");
+		} else {
+			LOG_WRN("nrf_modem_gnss_agps_expiry_get() failed, err %d", err);
+		}
+
+		/* Start and stop GNSS to trigger NRF_MODEM_GNSS_EVT_AGPS_REQ event if assistance
+		 * data is needed.
+		 */
+		nrf_modem_gnss_start();
+		nrf_modem_gnss_stop();
```

Comment:
Do we want to do these also in other than NRF_EOPNOTSUPP error cases? ENOMEM/ESHUTDOWN.

Reply:
That's probably not necessary. If the n`rf_modem_gnss_agps_expiry_get()` fails with `-NRF_ENOMEM` or `-NRF_ESHUTDOWN`, start/stop isn't most likely going to succeed either. I will change the implementation so, that start/stop is only done in case `-NRF_EOPNOTSUPP` is returned.